File: src/main/java/kr/co/lunatalk/domain/member/repository/MemberRepositoryImpl.java
package kr.co.lunatalk.domain.member.repository;

import com.querydsl.jpa.impl.JPAQueryFactory;
import kr.co.lunatalk.domain.member.domain.Member;
import kr.co.lunatalk.domain.member.domain.QMember;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

import static kr.co.lunatalk.domain.member.domain.QMember.*;

@Repository
@RequiredArgsConstructor
public class MemberRepositoryImpl implements MemberRepositoryCustom {
	private final JPAQueryFactory queryFactory;
	@Override
	public Page<Member> findMembers(Pageable pageable) {
		List<Member> content = queryFactory.selectFrom(member)
			.orderBy(member.createdAt.desc())
			.offset(pageable.getOffset())
			.limit(pageable.getPageSize())
			.fetch();


		Long total = Optional.ofNullable(
			queryFactory
				.select(member.count())
				.from(member)
				.fetchOne()
		).orElse(0L);

		System.out.println("OFFSET = " + pageable.getOffset());
		System.out.println("PAGE = " + pageable.getPageNumber());

		return new PageImpl<>(content,pageable, total);
	}
}

File: lunatalk-core/src/main/kotlin/kr/co/lunatalk/domain/member/repository/MemberRepositoryImpl.kt
package kr.co.lunatalk.domain.member.repository

import com.querydsl.jpa.impl.JPAQueryFactory
import kr.co.lunatalk.domain.member.domain.Member
import kr.co.lunatalk.domain.member.domain.QMember.member
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Repository

@Repository
class MemberRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : MemberRepositoryCustom {

    override fun findMembers(pageable: Pageable): Page<Member> {
        val content: List<Member> = queryFactory.selectFrom(member)
            .orderBy(member.createdAt.desc())
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())
            .fetch()

        val total: Long = queryFactory
            .select(member.count())
            .from(member)
            .fetchOne() ?: 0L

        return PageImpl(content, pageable, total)
    }
}
